README.md
# Go-Eat Compfest

This is a assignment for Compfest Software Engineering Academy Phase 2 Selection 2019. 

## Problem Description
You are going to make a command line app called “Go-Eat”. The app consists of drivers, stores, and a user. Every driver rated in the form of decimal number in the range of 0.0 to 5.0 and they will initially have 0 point of rating when the program runs for the first time. Each store has a menu which consists of item(s). The map which the user, drivers, and stores are placed on will be generated randomly and must be guaranteed that there is at least a path that can be taken for each driver to go to every store before finally proceeding to the user. Any additional logical assumptions is allowed.

## Your Go-Eat can be executed in three ways:
### 1. By running the app without any arguments
```
ruby go_eat.rb
```
### 2. By passing a filename as an argument
```
ruby go_eat.rb format.txt
```
#### Note the file format must following this format
```
mapsize {int value}
user_coordinate {int},{int}
drivers {int totalDriver} {name}-{coordinate X},{coordinate Y} {name}-{coordinate X},{coordinate Y} 
stores {int totalStore} {nameStore}-{coordinate X},{coordinate Y} {int totalMenu} {nameMenu}-{price} {nameMenu}-{price} {nameStore}-{coordinate X},{coordinate Y} {int totalMenu} {nameMenu}-{price} {nameMenu}-{price}
```

Example :
```
mapsize 20
user_coordinate 10,10
drivers 4 Alpha-11,15 Beta-2,7 Charlie-4,17 Echo-13,7
stores 2 warteg-12,13 4 Ikan_kembung-9 ayam_goreng-10 ayam_bakar-11 ayam_gulai-12 KFC-13,14 2 paha_atas-10 dada-11
```

### 3. By running the app with three arguments
```
ruby go_eat.rb 20 10 10
```
#### The user coordinate must be lower than the mapsize

## Once the app is running, user will be prompted choices of action.
### 1. Show Map

This is an example of 20x20 mapsize with 5 drivers, 3 stores, and 1 user.
```
| ....A...............
| ....................
| U'...................
| ....................
| ....................
| ....................
| ....................
| ........K'...........
| .......C............
| D...................
| ....................
| .........B..........
| .....E..............
| ....*...............
| ......W'.............
| ....................
| ....................
| ....................
| ....................
| ....................
------------------------
```
The user willl be denoted as *
The drivers will be denoted as the first letter in their name {A,B,C,D,E}
The stores will be denoted as the first letter in their name but added with aposthrope {K', W', U'}

### 2. Order Food
```
Choose the restaurant :
1. Warteg
2. Kfc
3. Upnormal
Insert the number >
```
User will choose the number that represent their chosen restaurant

```
Warteg Menu
Menu - Harga
Ayam - 10
Udang - 20
Kerang - 30
Masukkan menu yang anda pilih dengan format (Nama-Banyaknya)
Ketik 'OK' jika sudah selesai pesan dan 'Cancel' untuk mengosongkan pesanan
```
User will input the menu with format 
```Menu-Quantity```
and then type 'OK' if they already done with their order.

```
Your order
Menu - Harga - Quantity - Delivery fee
Ayam - 10 - 1 - 30
Udang - 20 - 1 - 60
Kerang - 30 - 1 - 90
Total Biaya : 240
Are you sure with your order? [Y/N]
```
Program will ask for order confirmation.

After type 'Y' when user already sure with the order.
```
Store name : Warteg
Store location : 7,6
Store distance : 3
Driver name : Echo
Driver rating : 0
Order :
Ayam - 10 - 1 - 30
Udang - 20 - 1 - 60
Kerang - 30 - 1 - 90
Total price : 240
Route >
Start at [7, 6]
- Go to [6, 6]
- Go to [5, 6]
- Go to [5, 7]
- Finish at [5, 7]
Enjoy your food
```
The program will generate the order information and the driver journey.

```Please give rating (0 - 5) :```

The program ask for your rating to the driver.

## Assumption Used

* The history will still after quit the program, but will reset when the program start
* The map coordinate start (0,0) at left-bottom and (maxsize,maxsize) at top-right
* The distance is based on vertical and horizontal move
* Nearest driver choosed from the vertical and horizontal step, not diagonal step
* If there are more than 1 nearest driver, the program will only choose 1
* Driver will shown in map by first letter in their name 
* Store will shown in map by first letter in their name with aposthrope (')
* User will shown in map by (*)


history.txt
{1
Order from Upnormal at [20, 16]
Driver's name : Charlie
Deliver to : [7, 10]
Item's ordered :
Menu - Harga - Quantity - Delivery fee
Mie_goreng - 12 -1 -228
Total price : 240
Route :
Start at [20, 16]
- Go to [19, 16]
- Go to [18, 16]
- Go to [17, 16]
- Go to [16, 16]
- Go to [15, 16]
- Go to [14, 16]
- Go to [13, 16]
- Go to [12, 16]
- Go to [11, 16]
- Go to [10, 16]
- Go to [9, 16]
- Go to [8, 16]
- Go to [7, 16]
- Go to [7, 15]
- Go to [7, 14]
- Go to [7, 13]
- Go to [7, 12]
- Go to [7, 11]
- Go to [7, 10]
- Finish at [7, 10]
Enjoy your food


map_generator.rb
require_relative "user"
require_relative "driver"
require_relative "store"
require_relative "coordinate"
require_relative "format"
require_relative "map"

$stores_name = [["Warteg", {"Ayam" => 10,"Udang" => 20,"Kerang" => 30}], ["KFC", {"Paha_atas" => 5, "Sayap" => 7, "Paha_bawah" => 9}], ["Upnormal", {"Mie_goreng" => 12, "Mie_rebus" => 15, "Mie_bakar" => 18}]]

class MapGenerator
    attr_accessor :map
    attr_accessor :drivers
    attr_accessor :user
    attr_accessor :stores

    def initialize
        @map = 0
        @drivers = 0
        @user = 0
        @stores = 0
    end

    def generate(args)
        if args.size == 0
            @drivers = drawRandomDrivers(5,20)
            @user = drawRandomUser(20)
            @stores = drawRandomStores(3, 20)

            gmap(20)
        
        elsif args.size==1
            data = Format.new(args[0])

            @drivers = data.drivers
            @user = data.user
            @stores = data.stores
            
            gmap(data.mapsize)
        
        elsif args.size==3
            @drivers=drawRandomDrivers(5,args[0].to_i)
            @stores = drawRandomStores(3, args[0].to_i)
            @user = User.new(Coordinate.new(args[1].to_i, args[2].to_i))

            gmap(args[0].to_i)
        end
    end

    def gmap(mapsize)
        @map = Map.new(mapsize)
        @map.addDrivers(@drivers)
        @map.addStores(@stores)
        @map.addUser(@user)
    end

    def drawRandomUser(mapsize)
        a = User.new(Coordinate.new(mapsize))
        a
    end

    def drawRandomDrivers(n, mapsize)
        drivers = {}
        while drivers.size < n
            to_driver = Driver.new(Coordinate.new(mapsize))
            if drivers.key?(to_driver.name) || drivers.has_value?([to_driver.x, to_driver.y])
                
            else
                drivers[to_driver.name] = [to_driver.x, to_driver.y]
            end
        end
        
        result = {}
        drivers.each do |key, val|
            result[key] = Driver.new(Coordinate.new(mapsize))
            result[key].coordinate = Coordinate.new(val[0], val[1])
            result[key].x = val[0]
            result[key].y = val[1]
            result[key].name = key
        end

        result.values
    end

    def drawRandomStores(n, mapsize)
        stores = {}
        i = 0
        while stores.size < n
            to_store = Store.new($stores_name[i][0].capitalize, Coordinate.new(mapsize))
            if stores.key?(to_store.name) || stores.has_value?([to_store.x, to_store.y])
                
            else
                stores[to_store.name] = [to_store.x, to_store.y]
            end
            i += 1
        end
        
        result = {}
        i = 0
        stores.each do |key, val|
            result[key] = Store.new($stores_name[i][0].capitalize, Coordinate.new(mapsize))
            result[key].generate_menu($stores_name[i][1])
            result[key].coordinate = Coordinate.new(val[0], val[1])
            result[key].x = val[0]
            result[key].y = val[1]
            result[key].name = key
            i += 1
        end
        result.values
    end 

    def getNearestDriver(store)
        nearestDriver = nil
        nearestDistance = 99999
        s = store.coordinate
        drivers.each do |d|
            distance = (s.x - d.coordinate.x).abs + (s.y - d.coordinate.y).abs
            if distance<nearestDistance
                nearestDistance = distance
                nearestDriver = d
            end
        end
        nearestDriver
    end

    def randomizeDriverFrom(order)
        intoHash = []
        theDriver = 0
        @drivers.each do |d|
            if d.name != order.driver.name
                intoHash.push(d.name)
                intoHash.push(d.coordinate.get)
            else
                theDriver = d
            end
        end
        drivers = HASH[intoHash.each_slice(2).to_a]

        sizeEst = drivers.size + 1
        while drivers.size != sizeEst
            newCoordinate = Coordinate.new(@map.height)
            theDriver.coordinate = newCoordinate
            theDriver.x = newCoordinate.x
            theDriver.y = newCoordinate.y

            if drivers.key?(theDriver.name) || drivers.has_value?([theDriver.x, theDriver.y])

            else 
                drivers[theDriver.name] = [theDriver.x, theDriver.y]
            end
        end

        result = {}
		drivers.each do |key, val| 
			result[key] = Driver.new(Coordinate.new(@map.height))
			result[key].coordinate = Coordinate.new(val[0],val[1])
			result[key].x = val[0]
			result[key].y = val[1]
			result[key].name = key 
        end
        
        result.values
    end

    def generateMapRoute(order)
        routeMap = @map
        routeMap.addRoute(order)
        routeMap
    end
end



